From e93e507d1ab1d16464e10763a9c6f33a73441974 Mon Sep 17 00:00:00 2001
From: Michel Becle <michelx.becle@intel.com>
Date: Thu, 25 Apr 2013 11:34:06 +0200
Subject: [PORT FROM MAIN] wifi: DUT won't reconnect after 2.4 / 5 Ghz band switch after WIFI on/off

BZ: 98644

During the process of switching from 2.4 Ghz to 5 Ghz bands and vice versa
the wpa supplicant is not informed of changes made by the user from the GUI
The only way to inform the wpa supplicant a band change has happenned is to read the
information back from the WIFI chip.
This patch introduces a refresh of the wpa_s->hw.modes structure on scan results event.
With the now up to date hw.mode data, the supplicant can reconnect on the active bands.

Change-Id: Ic5ba283716d6ab4062fd2bd530b67fa554d32ea1
Orig-Change-Id: I96857aaeed7b971abefb4dc6b9f59f91715837be
Category: device enablement
Domain: CWS.WIFI-Common
Origin: upstream
Upstream-Candidate: no,proprietary
Signed-off-by: Michel Becle <michelx.becle@intel.com>
Signed-off-by: Jean-Michel Bachot <jean-michel.bachot@intel.com>
---
 wpa_supplicant/bestchannel.c |    8 --------
 wpa_supplicant/events.c      |   13 +++++++++++++
 2 files changed, 13 insertions(+), 8 deletions(-)

diff --git a/wpa_supplicant/bestchannel.c b/wpa_supplicant/bestchannel.c
index b7e6082..094c8eb 100644
--- a/wpa_supplicant/bestchannel.c
+++ b/wpa_supplicant/bestchannel.c
@@ -183,14 +183,6 @@ int best_channel_get_freq_list(struct wpa_supplicant *wpa_s)
 	struct hostapd_hw_modes *modes;
 	int i, j;
 
-	/*
-	 * "wpa_s->hw" is not dynamically updated on a country code update.
-	 * So, refresh it before processing it in case country code has changed.
-	 */
-	wpa_s->hw.modes = wpa_drv_get_hw_feature_data(wpa_s,
-						      &wpa_s->hw.num_modes,
-						      &wpa_s->hw.flags);
-
 	modes = wpa_s->hw.modes;
 	if (modes == NULL)
 		return -1;
diff --git a/wpa_supplicant/events.c b/wpa_supplicant/events.c
index 1141509..000efbc 100644
--- a/wpa_supplicant/events.c
+++ b/wpa_supplicant/events.c
@@ -1159,6 +1159,19 @@ static int _wpa_supplicant_event_scan_results(struct wpa_supplicant *wpa_s,
 	static unsigned int best_chan_scan_cnt = 0;
 #endif
 
+	/*
+	 * "wpa_s->hw" is not dynamically updated on a country code update or RF setband change.
+	 * So, refresh it before processing it in case country or RF band code has changed.
+	 */
+
+	wpa_s->hw.modes = wpa_drv_get_hw_feature_data(wpa_s,
+						      &wpa_s->hw.num_modes,
+						      &wpa_s->hw.flags);
+
+#ifdef CONFIG_P2P
+	wpas_p2p_update_channel_list(wpa_s);
+#endif /* CONFIG_P2P */
+
 #ifdef CONFIG_AP
 	if (wpa_s->ap_iface)
 		ap = 1;
-- 
1.7.4.1

